networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_mgmt.py

```diff
@@ -0,0 +1,880 @@
+            "destinations": current if direction in 'ingress' else target,
+            "disabled": False,
+            "display_name": id,
+            "services": services,
+            "action": "ALLOW",
+            "logged": False, # TODO selective logging
+            "rule_tag": id.replace("-",""),
+            "_revision": provider_rule["_revision"]
+        }
+
+    def sg_rules_remote(self, cidr):
+        # NSX bug. Related IPSet to handle  0.0.0.0/x and ::0/x
+        return {
+            "resource_type": "IPSet",
+            "display_name": cidr,
+            "ip_addresses": [cidr],
+            "tags": self.tags(None)
+        }
+
+    def _sg_rule_target(self, os_rule, provider_rule):
+
+        if os_rule.get("remote_group_id"):
+            id = provider_rule.get("remote_group_id")
+            name = os_rule["remote_group_id"]
+            type = "IPSet"
+        elif provider_rule.get("remote_ip_prefix_id"):
+            # Non-OpenStack property "remote_ip_prefix_id"
+            # Used due to limitations of NSX-T Management API
+            id = provider_rule.get("remote_ip_prefix_id")
+            name = os_rule["remote_ip_prefix"]
+            type = "IPSet"
+        elif os_rule.get("remote_ip_prefix"):
+            id = name = str(netaddr.IPNetwork(os_rule["remote_ip_prefix"]))
+            type = {'IPv4': 'IPv4Address', 'IPv6': 'IPv6Address'}.get(os_rule['ethertype'])
+        else:
+            # Any
+            return []
+
+        return [{
+            "target_type": type,
+            "target_id": id,
+            "is_valid": True,
+            "target_display_name": name
+        }]
+
+
+    def _sg_rule_service(self, os_rule, provider_rule, subtype="NSService"):
+        min = os_rule["port_range_min"]
+        max = os_rule["port_range_max"]
+        protocol = os_rule["protocol"]
+        ethertype = os_rule['ethertype']
+
+        if protocol == 'icmp':
+            min = int(min) if str(min).isdigit() else min
+            max = int(max) if str(max).isdigit() else max
+
+            if min not in VALID_ICMP_RANGES[ethertype] or \
+                max not in VALID_ICMP_RANGES[ethertype][min]:
+                return \
+                    (None, "Not supported ICMP Range {}-{}".format(min, max))
+
+            return ({
+                "resource_type": "ICMPType{}".format(subtype),
+                "icmp_type": str(min),
+                "icmp_code": str(max),
+                "protocol": { 
+                    'IPv4': "ICMPv4", 
+                    'IPv6': "ICMPv6"
+                }.get(ethertype)
+            }, None)
+
+        if protocol in ["tcp", "udp"]:
+            return ({
+                "resource_type": "L4PortSet{}".format(subtype),
+                "l4_protocol": {'tcp': "TCP", 'udp': "UDP"}.get(protocol),
+                "destination_ports": ["{}-{}".format(min, max) \
+                    if min != max and max else str(min)],
```

Comment:
This causes an Error if an user creates rules with without min/max destination port or port = 0:
https://docs.openstack.org/api-ref/network/v2/?expanded=create-security-group-rule-detail#create-security-group-rule
Which causes min and max to be Null.

```python
  "details" : "Field level validation errors: {value 'None' of property rules[287].service_entries[0].destination_ports[0] violates format 'port-or-range'}",
  "httpStatus" : "BAD_REQUEST",
  "error_code" : 255,
  "module_name" : "common-services",
  "error_message" : "Field level validation errors: {value 'None' of property rules[287].service_entries[0].destination_ports[0] violates format 'port-or-range'}"
```

Possible fix:
```python
--- a/networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_mgmt.py
+++ b/networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_mgmt.py
@@ -440,11 +440,14 @@ class Payload(object):
             }, None)

         if protocol in ["tcp", "udp"]:
+            destination_ports = "1-65535"
+            if min or max:
+                destination_ports = "{}-{}".format(min, max) \
+                    if min != max and max else str(min)
             return ({
                 "resource_type": "L4PortSet{}".format(subtype),
                 "l4_protocol": {'tcp': "TCP", 'udp': "UDP"}.get(protocol),
-                "destination_ports": ["{}-{}".format(min, max) \
-                    if min != max and max else str(min)],
+                "destination_ports": [destination_ports],
                 "source_ports": ["1-65535"]
             }, None)
```

Reply:
Fixed.